core/src/main/scala/org/apache/spark/rdd/OrderedRDDFunctions.scala

```diff
@@ -72,6 +72,8 @@ class OrderedRDDFunctions[K : Ordering : ClassTag,
    * because it can push the sorting down into the shuffle machinery.
    */
   def repartitionAndSortWithinPartitions(partitioner: Partitioner): RDD[(K, V)] = {
+    val keyClass = implicitly[ClassTag[K]].runtimeClass
+    Partitioner.assertHashCodeIsWellBehaved(keyClass)
```

Comment:
I think the concern here is that there's other code which might call `hashCode` on the items, so even if we use a custom partitioner we will still get incorrect results for Java arrays.  I think that this is safe as long as we can guarantee that the sorting phase doesn't call `hashCode`.